src/corell/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+    TriangleList,
+    /// Every three consecutive vertices represent a single triangle. For example, with `[a, b, c,
+    /// d]`, `a`, `b`, and `c` form a triangle, and `b`, `c`, and `d` form a triangle.
+    TriangleStrip,
+    /// Patch list,
+    /// used with shaders capable of producing primitives on their own (tessellation)
+    PatchList(PatchSize),
+}
+
+/// A type of each index value in the slice's index buffer
+#[derive(Eq, Ord, PartialEq, PartialOrd, Hash, Copy, Clone, Debug)]
+#[allow(missing_docs)]
+#[repr(u8)]
+pub enum IndexType {
+    U16,
+    U32,
+}
+
+pub trait Instance {
+    type B: Backend;
+
+    fn create() -> Self;
+
+    // TODO: Use an iterator instead of Vec?
+    fn enumerate_physical_devices(&self) -> Vec<<<Self as Instance>::B as Backend>::PhysicalDevice>;
+}
+
+pub trait PhysicalDevice {
+    type B: Backend;
+
+    fn open(&self) -> (<<Self as PhysicalDevice>::B as Backend>::Device, Vec<<<Self as PhysicalDevice>::B as Backend>::CommandQueue>);
+    fn get_info(&self) -> PhysicalDeviceInfo;
+}
+
+#[derive(Clone, Debug)]
+pub struct PhysicalDeviceInfo {
+    /// Phyiscal device name
+    pub name: String,
+    /// Vendor PCI id of the physical device
+    pub vendor: usize,
+    /// PCI id of the physical device
+    pub device: usize,
+    /// The device is based on a software rasterizer
+    pub software_rendering: bool,
+}
+
+pub trait Device {
+    
+}
+
+pub trait CommandQueue {
+    type B: Backend;
+
+    /// Submits a `CommandBuffer` to the GPU queue for execution.
+    fn submit(&mut self, cmd_buffer: &<<Self as CommandQueue>::B as Backend>::CommandBuffer);
+}
+
+pub trait SwapChain {
+    type B: Backend;
+
+    fn present(&mut self);
+}
+
+/// Different resource types of a specific API. 
+pub trait Resources:          Clone + Hash + Debug + Eq + PartialEq + Any {
+    type Buffer:              Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync + Copy;
+    type Shader:              Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
+    type RenderPass:          Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
+    type PipelineLayout:      Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
+    type PipelineStateObject: Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
+    type Image:               Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
+    type ShaderResourceView:  Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync + Copy;
+    type UnorderedAccessView: Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync + Copy;
+    type RenderTargetView:    Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync + Copy;
+    type DepthStencilView:    Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
+    type Sampler:             Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync + Copy;
+}
+
+/// Different types of a specific API.
+pub trait Backend {
+    type CommandBuffer;
```

Comment:
if there is no `CommandBuffer` trait, perhaps we need some `RenderEncoder`, `TransferEncoder` and `ComputeEncoder` ones?